cou/apps/auxiliary.py

```diff
@@ -204,10 +205,55 @@ class MysqlInnodbCluster(AuxiliaryApplication):
     wait_timeout = LONG_IDLE_TIMEOUT
 
 
-# NOTE (gabrielcocenza): Although CephOSD class is empty now, it will be
-# necessary to add post upgrade plan to set require-osd-release. Registering on
-# AuxiliaryApplication can be easily forgot and ceph-osd can't be instantiated
-# as a normal OpenStackApplication.
 @AppFactory.register_application(["ceph-osd"])
 class CephOSD(AuxiliaryApplication):
     """Application for ceph-osd."""
+
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[COUUnit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[COUUnit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = [
+            PreUpgradeStep(
+                description="Check if colocated nova-compute units had been upgraded.",
+                coro=self._verify_nova_compute(target),
+            )
+        ]
+        steps.extend(super().pre_upgrade_steps(target, units))
+        return steps
+
+    async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
+        """Check if a nova-compute application has upgraded its workload version.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises ApplicationError: When any nova-compute app workload version isn't reached.
+        """
+        units_not_upgraded = []
+        apps = await self.model.get_applications()
+
+        for app in apps:
+            if app.charm != "nova-compute":
+                logger.debug("skipping application %s", app.name)
+                continue
+
+            for unit in app.units:
+                compatible_os_versions = OpenStackCodenameLookup.find_compatible_versions(
+                    app.charm, unit.workload_version
+                )
+
+                if target not in compatible_os_versions:
+                    units_not_upgraded.append(unit.name)
```

Comment:
```suggestion
            if len(self.os_release_units.keys()) > 1:
                units_not_upgraded = [units for os_version, units in  self.os_release_units.values if os_version != target]
                raise ApplicationError(
                f"Units '{', '.join(units_not_upgraded)}' did not reach {target}."
            )
            
```

Reply:
Actually I think it would be nicer to have a function like this on nova-compute:

```python
    def units_to_upgrade(self, target) -> list[COUUnit]:
        """Units that still need to upgrade to reach a target.

        :param target: OpenStack release as target to upgrade.
        :type target: OpenStack release.
        :return: list of units to upgrade
        :rtype: list[COUUnit]
        """
        return [
            units for units, os_version in self.os_release_units.items() if os_version < target
        ]
```

And then we could use that here like:
```python
units_not_upgraded = nova_compute.units_to_upgrade(target)
if units_not_upgraded:
    raise ApplicationError(...)
```

Reply:
Like this we will raise exception with not all information, if there are multiple nova-compute, we will check it only for single one.

Reply:
It's just the general idea. You could append units on `units_not_upgraded` and then use the same way.

Reply:
You are forgetting, that the app is COUApplication and not NovaCompute.

Reply:
True.. I forgot it :sweat_smile: . Solving it